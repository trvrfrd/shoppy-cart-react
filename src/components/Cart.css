:local(.cart) {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  margin-top: -10%;
  min-height: 50%;
  min-width: 50%;
}

:local(.headline) {
  text-align: center;
}
