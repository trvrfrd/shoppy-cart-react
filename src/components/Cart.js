import React, { Fragment } from 'react';
import CartTable from './CartTable';
import CartTotal from './CartTotal';
import CartControls from './CartControls';

import styles from './Cart.css';

export default function Cart({ items = [], onClear, onClose, onRemoveItem }) {
  return (
    <div className={styles.cart}>
      <h1 className={styles.headline}>Your Cart</h1>
      {items.length ?
        <Fragment>
          <CartTable
            items={items}
            onRemoveItem={onRemoveItem}
          />
          <CartTotal items={items} />
        </Fragment>
        : <p>Your cart is empty.</p>
      }
      <CartControls
        onClear={onClear}
        onClose={onClose}
      />
    </div>
  )
}
