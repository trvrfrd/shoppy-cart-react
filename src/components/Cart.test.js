import React from 'react';
import { mount } from 'enzyme';
import TestRenderer from 'react-test-renderer';
import Cart from './Cart';

const fakeItems = [
  {
    type: 'a',
    quantity: 1,
    price: 1.00
  },
  {
    type: 'b',
    quantity: 2,
    price: 1.50
  },
  {
    type: 'c',
    quantity: 3,
    price: 2.00
  }
];

describe('UI and rendering', () => {

  it('matches snapshot', () => {
    const wrapper = TestRenderer.create(<Cart items={fakeItems} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a CartTable with items as prop', () => {
    const wrapper = mount(<Cart items={fakeItems} />);
    const table = wrapper.find('CartTable');

    expect(table.prop('items')).toEqual(fakeItems);
  });

  it('renders a CartTotal with items as prop', () => {
    const wrapper = mount(<Cart items={fakeItems} />);
    const table = wrapper.find('CartTotal');

    expect(table.prop('items')).toEqual(fakeItems);
  });

  describe('controls', () => {

    it('renders CartControls', () => {
      const wrapper = mount(<Cart items={fakeItems} />);
      const controls = wrapper.find('CartControls');

      expect(controls.length).toBe(1);
    });

    it('has a Clear button that calls onClear prop', () => {
      const handler = jest.fn();
      const wrapper = mount(<Cart items={fakeItems} onClear={handler} />);
      const controls = wrapper.find('CartControls');
      const clear = controls.find('button').at(0);

      clear.simulate('click');

      expect(handler).toHaveBeenCalled();
    });

    it('has a Close button that calls onClose prop', () => {
      const handler = jest.fn();
      const wrapper = mount(<Cart items={fakeItems} onClose={handler} />);
      const controls = wrapper.find('CartControls');
      const close = controls.find('button').at(1);

      close.simulate('click');

      expect(handler).toHaveBeenCalled();
    });

    it('calls onRemoveItem with correct index when an item is removed', () => {
      const handler = jest.fn();
      const index = 0;
      const wrapper = mount(<Cart items={fakeItems} onRemoveItem={handler} />);
      const removeFirst = wrapper.find('[data-test-name="remove-item"]').at(index);

      removeFirst.simulate('click');

      expect(handler).toHaveBeenCalledWith(index);
    });
  });
});
