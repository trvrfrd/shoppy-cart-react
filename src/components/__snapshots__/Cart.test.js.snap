// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI and rendering matches snapshot 1`] = `
<div
  className="cart"
>
  <h1
    className="headline"
  >
    Your Cart
  </h1>
  <table
    className="table"
  >
    <thead
      className="header"
    >
      <tr>
        <th
          className="cell"
        >
          Item
        </th>
        <th
          className="cell"
        >
          Price
        </th>
        <th
          className="cell"
        >
          Qty
        </th>
        <th
          className="cell"
        />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="cell"
        >
          A
        </td>
        <td
          className="cell"
        >
          $1.00
        </td>
        <td
          className="cell"
        >
          1
        </td>
        <td
          className="cell"
        >
          <button
            className="button"
            data-test-name="remove-item"
            onClick={[Function]}
          >
            Remove
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="cell"
        >
          B
        </td>
        <td
          className="cell"
        >
          $1.50
        </td>
        <td
          className="cell"
        >
          2
        </td>
        <td
          className="cell"
        >
          <button
            className="button"
            data-test-name="remove-item"
            onClick={[Function]}
          >
            Remove
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="cell"
        >
          C
        </td>
        <td
          className="cell"
        >
          $2.00
        </td>
        <td
          className="cell"
        >
          3
        </td>
        <td
          className="cell"
        >
          <button
            className="button"
            data-test-name="remove-item"
            onClick={[Function]}
          >
            Remove
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <p
    className="total"
  >
    Total: 
    $10.00
  </p>
  <div
    className="wrapper"
  >
    <button
      className="button"
    >
      Clear
    </button>
    <button
      className="button"
    >
      Close
    </button>
  </div>
</div>
`;
